activemodel/lib/active_model/type/helpers/numeric.rb

```diff
@@ -25,10 +25,17 @@ def cast(value)
         end
 
         def changed?(old_value, _new_value, new_value_before_type_cast) # :nodoc:
-          super || number_to_non_number?(old_value, new_value_before_type_cast)
+          (super || number_to_non_number?(old_value, new_value_before_type_cast)) &&
+            !equal_nan?(old_value, new_value_before_type_cast)
         end
 
         private
+          def equal_nan?(old_value, new_value)
+            new_and_old_numeric = (old_value.is_a?(::Float) || old_value.is_a?(BigDecimal)) &&
+              (new_value.is_a?(::Float) || new_value.is_a?(BigDecimal))
+            new_and_old_numeric && (old_value.nan? && new_value.nan?) && old_value.instance_of?(new_value.class)
```

Comment:
This can be simplified further:
```suggestion
            (old_value.is_a?(::Float) || old_value.is_a?(BigDecimal)) &&
              old_value.nan? &&
              old_value.instance_of?(new_value.class) &&
              new_value.nan?
```

In theory `old_value.instance_of?(new_value.class)` could cause problem if `new_value` is a subclass of `Float` or something like that, but we're really going into weird edge cases.

Reply:
Yeah I thinked about that possibility but I think that would be really rare such case happen still being possible. This scenario with NaN I think is not much common too.